File: backend/theatermgnt/src/main/java/com/theatermgnt/theatermgnt/movie/entity/Movie.java
package com.theatermgnt.theatermgnt.movie.entity;

import java.time.LocalDate;
import java.util.HashSet;
import java.util.Set;

import jakarta.persistence.*;

import com.theatermgnt.theatermgnt.common.entity.BaseEntity;
import com.theatermgnt.theatermgnt.common.enums.MovieStatus;

import lombok.*;
import lombok.experimental.FieldDefaults;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@FieldDefaults(level = AccessLevel.PRIVATE)
@Entity
@Table(name = "movies")
public class Movie extends BaseEntity {
    String title;
    String description;
    Integer durationMinutes;
    String director;

    String castMembers;

    String posterUrl;
    String trailerUrl;
    LocalDate releaseDate;
    LocalDate endDate;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "age_rating_id")
    AgeRating ageRating;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", length = 20)
    MovieStatus status;

    @ManyToMany(
            fetch = FetchType.LAZY,
            cascade = {CascadeType.PERSIST, CascadeType.MERGE})
    @JoinTable(
            name = "movie_genres",
            joinColumns = @JoinColumn(name = "movie_id"),
            inverseJoinColumns = @JoinColumn(name = "genre_id"))
    Set<Genre> genres = new HashSet<>();

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Movie)) return false;
        Movie movie = (Movie) o;
        return id != null && id.equals(movie.id);
    }

    @Override
    public int hashCode() {
        return getClass().hashCode();
    }
}

File: frontend/admin/src/types/MovieType/Movie.ts
export interface AgeRating {
  id: string;
  code: string;
  description: string;
}

export interface Genre {
  id: string;
  name: string;
}

// Simple response for list view
export interface MovieSimple {
  id: string;
  title: string;
  posterUrl: string;
  durationMinutes: number;
  releaseDate: string;
  status: "now_showing" | "coming_soon" | "archived";
  ageRatingCode: string;
  director: string;
}

// Full response for detail view
export interface Movie {
  id: string;
  title: string;
  description: string;
  durationMinutes: number;
  director: string;
  cast: string;
  posterUrl: string;
  trailerUrl: string;
  releaseDate: string;
  endDate: string;
  status: "now_showing" | "coming_soon" | "archived";
  ageRating: AgeRating;
  genres: Genre[];
  createdAt?: string;
  updatedAt?: string;
}

export interface CreateMovieRequest {
  title: string;
  description: string;
  durationMinutes: number;
  director: string;
  cast: string;
  posterUrl: string;
  trailerUrl: string;
  releaseDate: string;
  endDate: string;
  status: "now_showing" | "coming_soon" | "archived";
  ageRatingId: string;
  genreIds: string[];
}

export interface UpdateMovieRequest {
  title?: string;
  description?: string;
  durationMinutes?: number;
  director?: string;
  cast?: string;
  posterUrl?: string;
  trailerUrl?: string;
  releaseDate?: string;
  endDate?: string;
  status?: "now_showing" | "coming_soon" | "archived";
  ageRatingId?: string;
  genreIds?: string[];
}
